Builtin: Print warnings if SHA256 of files does not match with build

Desour opened the pull request:
* Adds a list of builtin lua files in cmake. (like in src) (unittest files are not included)
* Generates a cpp file with relative paths and hashes.
* Prints warnings if you try to `loadfile` (or `dofile`) a path in builtin (with abs path) if either the file is not in the cmake list or if the hash mismatches.
  Fixes #16318.
* Code cleanup of `ScriptApiSecurity::safeLoadFile`.
  (And CPCSM files will now have the shebang ignored (but it looks like luajit already does that).)
* (Will also be useful for SSCSM client builtin (check to prevent cheating) and SSCSM server builtin (to find out which files to send).)
* Doesn't work for CPCSM.
* I've decided for warnings and not errors, as I'm worried that otherwise the user might be left without a working main menu (which right now results in a "No future without main menu!" plus abort) and not know what is happening.
* Side effect: The feature to load code from stdin with `dofile()` and `loadfile()` (without args) is removed.
  (ContendDB grep had no true positive results.)

## To do

This PR is Ready for Review.

TODO:
* [x] Idk how to properly escape paths in cmake. I.e. if a `"` is in a path, it *might* split the list. (see github review comment below)
* [x] On macOs there are issues with the file being added to multiple targets. TODO 

## How to test

* Start and notice how there are no warnings.
* Change a builtin file, e.g. after.lua. Observe the warning.
* Rebuild, and observe that it regenerates the hashes.
* Comment out a file in CMakeLists.txt. Observe the other warning.
* `/lua dofile()`
* Look into the generated builtin_files.cpp.

Desour committed:
tmp

Desour committed:
clean up a bit

Desour committed:
refactor safeLoadFile

Desour committed:
do hash checks

TODOs:
* move it somewhere else
* abs path
* mod security off codepath
* exception instead of returning false
* fallback mainmenu

Desour committed:
add all the files

(a bunch of `ls -1`)

Desour committed:
try to fix CI

(should maybe not use add_custom_command then?)

Desour committed:
try to fix server-only builds

Desour committed:
tmp

Desour committed:
(trivial) add a const

Desour committed:
move shebang check into a new safeLoadFileContent

(Should make shebangs work in CPCSM. (Though it looks like luajit already
ignores shebangs. :shrug:))

Desour committed:
Just print a warning on hash mismatch

Desour committed:
use abs paths for comparison

Desour committed:
remove hash param

Desour committed:
disallow stdin

Desour committed:
simplify file reading code

Desour committed:
oopsi, remove include

Desour committed:
remove old func

Desour committed:
remove unnecessary include

Desour committed:
dont need to pulbicate FileUniquePtr anymore

Desour committed:
use global var instead of fun

Desour committed:
try to fix macos

this should make GenerateBuiltinFilesCpp a common dependency of EngineCommon,
I think, so fix this:

>  The custom command generating
>
>    /Users/runner/work/luanti/luanti/build_xcode/src/builtin_files.cpp
>
>  is attached to multiple targets:
>
>    GenerateBuiltinFilesCpp
>    EngineCommon
>
>  but none of these is a common dependency of the other(s).  This is not
>  allowed by the Xcode "new build system".

Desour committed:
github, would you please take the commits into the PR?

Desour committed:
try with more *actual* force

Desour committed:
make BUILTIN_SRCS relative

Desour committed:
remove debug prints

I hate cmake.

Desour committed:
use relative paths in GenerateBuiltinFilesList.cmake

also makes GenerateBuiltinFilesList.cmake simpler

Desour committed:
don't add / to BUILTIN_BASE_PATH

Desour committed:
try to get rid of GenerateBuiltinFilesCpp

it generates a cpp file, so no src file will include it. hence clang tidy *should* not
need it, right?

Desour committed:
Revert "try to get rid of GenerateBuiltinFilesCpp"

This reverts commit ef6bc53e565387ed80d34e8a60b7dbf94489f44d.

the clang tidy ci does need it

Desour committed:
add new builtin file to cmakelist

Desour commented:
Rebased.

sfan5 commented on src/CMakeLists.txt:
```diff
@@ -395,6 +395,26 @@ add_custom_target(GenerateVersion
+	DEPENDS
+		${BUILTIN_SRCS_ABS}
+		"${PROJECT_SOURCE_DIR}/builtin_files.cpp.in"
+		"${CMAKE_SOURCE_DIR}/cmake/Modules/GenerateBuiltinFilesList.cmake"
```
nitpick but I'd call just "HashBuiltinFiles.cmake"

Desour committed:
use string views

Desour committed:
rename to HashBuiltinFiles.cmake

Desour committed:
fix comment

Desour committed:
use fs::PathStartsWith

sfan5 commented on src/builtin_files.cpp.in:
```diff
@@ -0,0 +1,13 @@
+
+using namespace std::literals::string_view_literals;
+
+extern const std::unordered_map<std::string_view, std::string_view> g_builtin_file_sha256_map = {
```
```suggestion
const std::unordered_map<std::string_view, std::string_view> g_builtin_file_sha256_map = {
```

Desour committed:
tmp

Desour committed:
ah yes, python exception in C++

    __os=..., __str=Python Exception <class 'OverflowError'>: int too big to convert
)
    at /usr/bin/../lib64/gcc/x86_64-pc-linux-gnu/15.2.1/../../../../include/c++/15.2.1/bits/basic_string.h:4372

Desour committed:
fix stuff, I think

Desour committed:
remove optional formatting again

Desour committed:
remove additional extern

I'd prefer with extern, but *shrug*

Desour commented on src/script/cpp_api/s_security.cpp:
```diff
@@ -457,72 +460,66 @@ bool ScriptApiSecurity::safeLoadString(lua_State *L, std::string_view code, cons
+		if (std::string_view(path_abs).substr(0, builtin_path_abs.size()) != builtin_path_abs)
+			break; // not in builtin
+
+		auto path_local = std::string_view(path_abs).substr(builtin_path_abs.size());
```
Done.

sfan5 commented on src/filesys.cpp:
```diff
@@ -714,6 +714,26 @@ bool PathStartsWith(const std::string &path, const std::string &prefix)
+
+	if (child_abs.size() == parent_abs.size()) {
+		assert(child_abs == parent_abs);
+		return "";
```
while technically correct I think it's potentially error-prone to have `""` be a valid path.
I would have this path return `"."` and reserve `""` for errors.

Desour committed:
dont use optional

:shrug:

Desour committed:
easy to use UASSERTEQ now

sfan5 commented on src/filesys.cpp:
```diff
@@ -714,6 +714,26 @@ bool PathStartsWith(const std::string &path, const std::string &prefix)
+		return "";
+	} else {
+		assert(child_abs.size() >= parent_abs.size() + 1);
+		assert(child_abs[parent_abs.size()] == DIR_DELIM_CHAR);
```
if it's not specified explicitly then I think we shouldn't rely on it.

Desour replied:
TL;DR: Looked again. It is (handwavingly) specified.

==> We can rely on it imo.

I've added unit test cases for `AbsolutePath` accordingly.

---

I've searched for more documentation. Unfortunately, I haven't found any kind of specification for the win32 api, only the doc where they want to present me machine translated german and burry the language option at the bottom, and exact details are so spare.

For comparison, the `realpath()` quote is literally the start of the first sentence of the man page.

Anyways, I've found this:
https://learn.microsoft.com/en-us/dotnet/standard/io/file-path-formats#path-normalization

>Almost all paths passed to Windows APIs are normalized.
>[...]
>This normalization happens implicitly, but you can do it explicitly by calling the [Path.GetFullPath](https://learn.microsoft.com/en-us/dotnet/api/system.io.path.getfullpath) method

It's probably fine to assume `_fullpath` also implicitly normalizes paths, also note the similar name to `Path.GetFullPath`.

>During normalization, Windows performs the following steps:
>[...]
> * Canonicalizes component and directory separators.
> * Evaluates relative directory components (. for the current directory and .. for the parent directory).
> * Trims certain characters.
>
>[...]
>### Canonicalize separators
>
>All forward slashes (`/`) are converted into the standard Windows separator, the back slash (`\`). If they are present, a series of slashes that follow the first two slashes are collapsed into a single slash.

The "first two slashes" probably refer to UNC paths.
The series of slashes "following" the first two *probably* doesn't mean that it has to be directly following, as the following section also speaks of runs of separators being removed:

>### Trim characters
>
>Along with the runs of separators and relative segments removed earlier,

Desour committed:
add abs path unittests

Desour committed:
add new vector2.lua to cmakelists

sfan5 commented on src/script/cpp_api/s_security.cpp:
```diff
@@ -662,72 +665,60 @@ bool ScriptApiSecurity::safeLoadString(lua_State *L, std::string_view code, cons
+		if (it->second != digest) {
+			warningstream << "SHA256 of builtin file \"" << path
+					<< "\" does not match. Expected: " << it->second
+					<< " Found: " << digest.c_str() << std::endl;;
```
```suggestion
					<< "\" does not match.\nExpected: " << it->second
					<< "\nFound   : " << digest << std::endl;
```

sfan5 commented on src/filesys.h:
```diff
@@ -111,6 +111,14 @@ bool MoveDir(const std::string &source, const std::string &target);
 bool PathStartsWith(const std::string &path, const std::string &prefix);
 
+// If child is (as absolute path) inside parent (also as absolute path), returns
+// the part of child that is relative to parent (without leading /).
```
nitpick but a relative path can't start with `/` so this is obvious
```suggestion
// the part of child that is relative to parent.
```

sfan5 commented on src/filesys.h:
```diff
@@ -111,6 +111,14 @@ bool MoveDir(const std::string &source, const std::string &target);
 
+// If child is (as absolute path) inside parent (also as absolute path), returns
+// the part of child that is relative to parent (without leading /).
+// Symlinks and "." and ".." components are removed.
```
this isn't exactly true due to the use of `AbsolutePathPartial`, right?

Desour replied:
What part do you mean exactly? The . and .. or symlinks?

`AbsolutePathPartial` removes symlinks at existing part of path. If a path component doesn't exist, it can't be a symlink, or can it?

. and .. are removed entirely.

sfan5 reviewed (commented):
Tested, works otherwise.

Desour committed:
some trivial text changes

sfan5 commented on src/unittest/test_filesys.cpp:
```diff
@@ -279,6 +340,9 @@ void TestFileSys::testAbsolutePath()
 		UASSERTEQ(auto, fs::AbsolutePath(dir_path2 + DIR_DELIM ".."), fs::AbsolutePath(dir_path));
+		// excess . and / are removed
+		UASSERTEQ(auto, fs::AbsolutePath(dir_path2 + p("//..")), fs::AbsolutePath(dir_path));
+		UASSERTEQ(auto, fs::AbsolutePath(dir_path2 + p("/./.././//")), fs::AbsolutePath(dir_path));
```
this fails on msvc btw, from CI:
```
Test assertion failed: fs::AbsolutePath(dir_path2 + p("/./.././//")) == fs::AbsolutePath(dir_path)
    actual  : C:\Users\RUNNER~1\AppData\Local\Temp\MT_a08080\
    expected: C:\Users\RUNNER~1\AppData\Local\Temp\MT_a08080
    at test_filesys.cpp:345
[FAIL] testAbsolutePath - 0ms
```

Desour replied:
testMakePathRelativeTo fails as well:
```
Caught unhandled exception: invalid string position
[FAIL] testMakePathRelativeTo - 3ms
```

Why doesn't the CI fail when its unittests fail?

sfan5 replied:
some tests are known to fail so it ignores the exit code entirely :man_shrugging:

Desour replied:
Tried to fix. Will see results in an hour:
https://github.com/Desour/minetest/actions/runs/27835241113/job/82381576080
